[TASK] Improve wording and add some additions to migration (#407)
TYPO3-Documentation/TYPO3CMS-Guide-HowToDocument: Documentation/Migration/Index.rst

    Text rendering to try it out. The new rendering will become mandatory in
    August 2024.

The main difference compared to the Sphinx rendering is that the PHP-based rendering
requires a file called :file:`Documentation/guides.xml` for configuration.
The Sphinx rendering required a file called :file:`Documentation/Settings.cfg`.

> In the transition period the process detects if a file called

Reviewer: Actually, I don't like the word "process" here. It sounds like you can run the guides rendering locally also for Settings.cfg, which is not the case. It is only related to the official rendering with our GitHub action.

Reviewer 2: How about my suggestions or what would you suggest?

> In the transition period the process detects if a file called

Reviewer 2: ```suggestion
In the transition period the GitHub action of the official rendering process detects if a file called
```

The last parameter (:bash:`Documentation`) is the name of the directory, where your
:file:`Settings.cfg` is currently placed in.

After the migration is performed, you will see some output in the terminal about which settings were
converted, if some old settings were discarded, or errors occurred. When you see this:

    ..  code-block:: text
        :caption: Output of the command

        Note: Some of your settings could not be converted:
            * html_theme_options
            * project_discussions
            * use_opensearch

everything went well. They can be ignored since these files are usually files that

> not need to be converted. You now can safely delete :file:`Settings.cfg`. Also delete

Reviewer: "You now can safely ..." sounds strange.

> not need to be converted. You now can safely delete :file:`Settings.cfg`. Also delete

Reviewer 2: ```suggestion
not need to be converted. You can now delete :file:`Settings.cfg`. If you ever should need it again it is still in the git history. Also delete
```

..  rst-class:: bignums

#.  Manual modifications of the :file:`guides.yml`

>     You have to manually change the following: in the extension tag add the attribute

Reviewer: ```suggestion
    You have to manually change the following: in the :xml:`<extension>` tag add the attribute
```

        ..  code-block:: text
            :caption: Changes in your :file:`guides.yml`

            interlink-shortcode="vendor/ext_name"

>     This you can find in your composer.json under "name". E.g. :json:`friendsoftypo3/jumpurl`.

Reviewer: "This you can find ...." sounds strange.

>     This you can find in your composer.json under "name". E.g. :json:`friendsoftypo3/jumpurl`.

Reviewer: ```suggestion
    This you can find in your :file:`composer.json` under "name". E.g. :json:`friendsoftypo3/jumpurl`.
```

>     We recommend to reformulate the code using the mac short cut `cmd + option/alt + l` and to

Reviewer: What about Windows/Linux users? No Max shortcut? Additionally, use the :kbd: role.

Additional "reformulare" sounds strange, better "reformat"

Author: @brotkrueml Suggest a nice sentence then I commit it.

    make docs

You should see something like this

.. code-block:: text

    Successfully placed 7 rendered HTML, SINGLEPAGE, and INTERLINK files into /project/Documentation-GENERATED-temp

..  _migrate-possible-errors:

Possible errors using `make docs`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

We provide four example errors to guide you through the fixes.

..  _migrate-interlink-inventory-not-found:

Interlink inventory not found: HTTP/2 404
"""""""""""""""""""""""""""""""""""""""""

.. code-block:: text

    [2024-03-13T12:22:50.661532+00:00] app.WARNING: Interlink inventory not found: HTTP/2 404
    returned for "https://docs.typo3.org/m/typo3/book-extbasefluid/11.5/en-us/objects.inv.json". [] []

Here you see that the link `https://docs.typo3.org/m/typo3/book-extbasefluid/11.5/en-us` is not found (404 page).

> We can now check via google if there is another link to book Extbasefluid. We found this site `https://docs.typo3.org/m/typo3/book-extbasefluid/10.4/en-us/`

Reviewer: ```suggestion
We can now check via Google if there is another link to book Extbasefluid. We found this site `https://docs.typo3.org/m/typo3/book-extbasefluid/10.4/en-us/`
```

> We can find the hint: `This manual is no longer being maintained for TYPO3 versions 11.5 and above.`. This tells us that the developers stopped to update the

Reviewer: ```suggestion
We can find the hint: `This manual is no longer being maintained for TYPO3 versions 11.5 and above.`. This tells us that the Documentation Team abandoned this
```

> documentation. We can for example link to the last existing version. Which is 10.4. To do this we have to change the :file:`guides.xml`. Search for the

Reviewer: ```suggestion
manual. We can, for example, link to the last existing version. Which is 10.4. To do this we have to change the :file:`guides.xml`. Search for the
```

Use the outdated 10.4 is not recommended. Instead, the content should be searched in the other manuals, mainly TYPO3 Explained.

Reviewer 2: When you are in a process of migrating the old documentation this is what you would do. And then recommend to replace the links in follow up work?

Author: When I remember correctly there was no other option than linking to the old documentation. I am not sure if linking to another article would be better here. This decision is best addressed to the owner of the extension builder.

Reviewer 2: @brotkrueml can you try to improve this section in a follow up please. This is beyond the scope of this PR

Reviewer: Personally, I would remove this section from here as this has nothing to do with the migration. Maybe move it later to a "Troubleshooting" section which is independent from the migration (and can also be linked from here). We should distinct between warnings appearing because of the migration and warning appearing for other reasons.

I may assume that people don't look into the migration chapter when warning appear in a year or two, but maybe into a troubleshooting chapter.

But fine for me for now, not blocking it.

.. code-block:: xml

    <inventory id="t3extbasebook"
           url="https://docs.typo3.org/m/typo3/book-extbasefluid/11.5/en-us/"
    />

and change `11.5` to `10.4`. Generate files again.

..  _migrate-inventory-link-with-key-not-found:

Inventory link with key ... not found
"""""""""""""""""""""""""""""""""""""

.. code-block:: text

    [2024-03-13T12:26:40.940930+00:00] app.WARNING: Inventory link with key "h2document:rest-common-pitfalls"
    (rest-common-pitfalls) not found.  {"rst-file":"GeneratedExtension/Index","type":"ref","targetRef

We see already that we have to go to file :file:`GeneratedExtension/Index` in the directory "Documentation".

> In there we have to delete the line which contains

Reviewer: I don't think deleting is always the best case. Better would be to look for the content in the same manual, but maybe another anchor.

Reviewer 2: In this case the whole section was removed. It depends on the use case what the best solution looks like here...

Reviewer: That is what I mean: You present a recipe, users follow it (maybe without thinking about it), which is maybe not what would be the best case.

Author: Me as a user wants to have at least a small hint what can be the correct behavior than having nothing at hand. I feel not professional enough to improve this article to satisfy your thoughts. So maybe another person can overtake this task.

Reviewer 2: @brotkrueml can you try to improve this section in a follow up please. This is beyond the scope of this PR

.. code-block:: text

    * :ref:`h2document:rest-common-pitfalls`

..  _migrate-nested-php-domain-components-not-supported:

Nested PHP domain components (php:class, php:interface, php:enum etc) are not supported
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

.. code-block:: text

    [2024-03-25T13:26:11.600367+00:00] app.WARNING: Nested PHP domain components
    (php:class, php:interface, php:enum etc) are not supported.
    Found php:\Vendor\MyExtension\Interfaces\RequireJsModuleInterface inside
    \Vendor\MyExtension\Interfaces\AnotherImportantInterface {"rst-file":"Developer/Index.rst"} []

> The file :file:`Index.rst` in Documentation/Developer has a wrong indentation. A class must not belong to another class.

Reviewer: ```suggestion
The file :file:`Index.rst` in :file:`Documentation/Developer/` has a wrong indentation. A class must not belong to another class.
```

We always append a slash at the end of folders to distinguish between folders and files.

Here is the wrong rst code.

.. code-block:: rst

    ..  php:class:: AnotherImportantInterface

    Used for ...

    ..  php:class:: RequireJsModuleInterface

    Widgets implementing this interface will add the provided RequireJS modules.
    Those modules will be loaded in dashboard view if the widget is added at least once.

Additionally regarding the name it has to be interfaces and not classes.
`..  php:class:: ExampleInterface` has to be changed to `..  php:interface:: ExampleInterface`.

..  _migrate-reference-could-not-be-resolved:

Reference sitehandling-addinglanguages could not be resolved in LocalizedContent/Index {"rst-file":"LocalizedContent/Index"} []
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

> The next step is to visit the old site, e.g. `https://docs.typo3.org/m/typo3/guide-frontendlocalization/main/en-us/LocalizedContent/Index.html`.

Reviewer: Why **old** site?

Reviewer 2: Because you can see if the refernce was still working in the old rendering?

Author: By "old" I meant the old links when using the Sphinx Rendering. As "new" I would refer to the PHP-based rendering.

Reviewer: But how can a user visit an "old site" when it is already rendered with guides?

As you see, I haven't understand it, so maybe our readers also don't.

There we can search for `sitehandling-addinglanguages` in the rst code by clicking the button "`</>View Source`".
Finding this:

.. code-block:: rst

    ..  tip::

>     For more information on how to add languages and configure their

Reviewer: Wrong indentation.

>     ..  tip::
>     For more information on how to add languages and configure their
>     behaviour in the site configuration, see
>     :ref:`Adding Languages <sitehandling-addinglanguages>`.

Reviewer 2: ```suggestion
    ..  tip::
        For more information on how to add languages and configure their
        behaviour in the site configuration, see
        :ref:`Adding Languages <sitehandling-addinglanguages>`.
```

    />

and change `11.5` to `10.4`. Generate files again.

..  _migrate-inventory-link-with-key-not-found:

Inventory link with key ... not found
"""""""""""""""""""""""""""""""""""""

.. code-block:: text

    [2024-03-13T12:26:40.940930+00:00] app.WARNING: Inventory link with key "h2document:rest-common-pitfalls"
    (rest-common-pitfalls) not found.  {"rst-file":"GeneratedExtension/Index","type":"ref","targetRef

We see already that we have to go to file :file:`GeneratedExtension/Index` in the directory "Documentation".
In there we have to delete the line which contains

.. code-block:: text

    * :ref:`h2document:rest-common-pitfalls`

But whether we should delete the reference or not depends on the special case. In this case the whole section was removed.
So here it is useful to delete the reference.
We conclude: In general it depends on the case itself what the best solution is.

..  _migrate-nested-php-domain-components-not-supported:

Nested PHP domain components (php:class, php:interface, php:enum etc) are not supported
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

.. code-block:: text

>     [2024-03-25T13:26:11.600367+00:00] app.WARNING: Nested PHP domain components

Reviewer: Actually, I don't understand the whole text of the warning. But this is another story.

But please use "MyVendor" instead of "Vendor" in the examples.

Author: I do not think that it is a good practice to manipulate warnings. I mean it came up like this, why I should change the "Vendor" name to "MyVendor?". In the other examples I agree but here I disagree.

Reviewer: It is an example for an error message. We do not provide the original source, do we? So this can be adjusted. But I don't insist for change, but like it for consistency reasons.

Reviewer 2: @brotkrueml can you try to improve this section in a follow up please. This is beyond the scope of this PR

    (php:class, php:interface, php:enum etc) are not supported.
    Found php:\Vendor\MyExtension\Interfaces\RequireJsModuleInterface inside
    \Vendor\MyExtension\Interfaces\AnotherImportantInterface {"rst-file":"Developer/Index.rst"} []

The file :file:`Index.rst` in :file:`Documentation/Developer/` has a wrong indentation. A class must not belong to another class.
Here is the wrong rst code.

> .. code-block:: rst

Reviewer: What is wrong in this code? I see no wrong indentation?

Reviewer 2: Because "A class must not belong to another class." See sentence above

    ..  php:class:: AnotherImportantInterface

    Used for ...

    ..  php:class:: RequireJsModuleInterface

    Widgets implementing this interface will add the provided RequireJS modules.
    Those modules will be loaded in dashboard view if the widget is added at least once.

Additionally regarding the name it has to be interfaces and not classes.

> `..  php:class:: ExampleInterface` has to be changed to `..  php:interface:: ExampleInterface`.

Reviewer: ```suggestion
:rst:`..  php:class:: ExampleInterface` has to be changed to :rst:`..  php:interface:: ExampleInterface`.
```

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

The next step is to visit the site which was rendered with the Sphinx rendering, in our example
`https://docs.typo3.org/m/typo3/guide-frontendlocalization/main/en-us/LocalizedContent/Index.html`.
There we can search for `sitehandling-addinglanguages` in the restructured text code by clicking the button "`</> View Source`".
We have found this:

.. code-block:: rst

    ..  tip::
        For more information on how to add languages and configure their
        behaviour in the site configuration, see
        :ref:`Adding Languages <sitehandling-addinglanguages>`.

The link is leading to
`https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ApiOverview/SiteHandling/AddLanguages.html#sitehandling-addinglanguages`.
We have to click the symbol next to the heading and copy the correct link
which is the one for restructured text

.. image:: /Images/get_link.png
    :class: with-shadow
    :width: 600px

.. code-block:: rst

    :ref:`Adding Languages <t3coreapi:sitehandling-addingLanguages>`

> You have to replace the correct link in e.g. :file:`Documentation/LocalizedContent/Index.rst` to

Reviewer: ```suggestion
You have to replace the correct link in, for example, :file:`Documentation/LocalizedContent/Index.rst` to
```

fix the error. Note: Since we already migrated the Frontend Localization Guide
(`https://docs.typo3.org/m/typo3/guide-frontendlocalization/main/en-us/LocalizedContent/Index.html`)

> to use the PHP based rendering you cannot find the state that we have shown above anymore.

Reviewer: ```suggestion
to use the PHP-based rendering you cannot find the state that we have shown above anymore.
```
